### ISSUE TITLE
compile_commands.json incomplete with multi-executable project

### ISSUE THREAD

**Issue by gentooise** [microsoft/vscode-cmake-tools#1132]
### Brief Issue Summary
I have a folder opened in VSCode, with a hierarchy of CMakeLists.txt (where the root one opens the others using `add_subdirectory` directives, and so on).
These CMakeLists hierarchy defines multiple targets (i.e. multiple `add_executable` directives), e.g. a bunch of unit tests.
CMake Tools is all set and I can see all the targets:
![image](https://user-images.githubusercontent.com/1735060/77057128-c60d5580-69d3-11ea-907e-bab61d1c59f6.png)

My issue is: why `compile_commands.json` does not contain all cpp files built by `all` target? Is it possible to do that?

### Expected:
In compile_commands.json I expect to have all cpp files used in any of the build targets.

### Apparent Behavior:
Some cpp files are missing from `compile_commands.json`: apparently the JSON contains cpp files of just one target.

![image](https://user-images.githubusercontent.com/1735060/77057236-f48b3080-69d3-11ea-9315-ebf21f04b7fc.png)

### Platform and Versions

- **Operating System**: Lubuntu 18.04
- **CMake Version**: cmake version 3.10.2
- **VSCode Version**: 1.43.0
- **CMake Tools Extension Version**: 1.3.1
- **Compiler/Toolchain**: GCC for x86_64-pokysdk-linux (MIPS cross-compiler mipsel-poky-linux-musl-gcc (GCC) 5.4.0)

**Comment by bobbrow** [comment]
Are you generating the compile_commands.json so that you can get IntelliSense in cpptools?  If you haven't tried it already, I would recommend using the configuration provider instead.  It replaces compile_commands.json and should contain more data that cpptools can use.

Run the "C/C++: Change Configuration Provider..." command and select "CMake Tools" to try it out.

**Comment by gentooise** [comment]
> Are you generating the compile_commands.json so that you can get IntelliSense in cpptools?

Yes! But it seems auto-generated anyway... and my question remains :)
Just for sake of curiosity, how is it generated? It considers just one of the executable targets? (the first one, the last one?). I think it would be good to know (I see very few references to `compile_commands.json` in the docs).

> If you haven't tried it already, I would recommend using the configuration provider instead. It replaces compile_commands.json and should contain more data that cpptools can use.
> 
> Run the "C/C++: Change Configuration Provider..." command and select "CMake Tools" to try it out.

Actually, my current configuration (.vscode/settings.json) is this one:

```
{
    "cmake.autoRestartBuild": true,
    "cmake.buildBeforeRun": false,
    "cmake.sourceDirectory": "${workspaceRoot}",
    "cmake.buildDirectory": "${workspaceRoot}/build",
//    "cmake.loggingLevel": "trace",

    "C_Cpp.loggingLevel": "Debug",
    "C_Cpp.updateChannel": "Insiders",
    "C_Cpp.default.cStandard": "c99",
    "C_Cpp.default.cppStandard": "c++03",
    "C_Cpp.default.intelliSenseMode": "gcc-x86",
    "C_Cpp.default.configurationProvider": "vector-of-bool.cmake-tools",
    "C_Cpp.default.systemIncludePath": [
        "/opt/dl-distro-tiny/2.2.4/sysroots/x86_64-pokysdk-linux/usr/lib/mipsel-poky-linux/gcc/mipsel-poky-linux/5.4.0/include",
        "/opt/dl-distro-tiny/2.2.4/sysroots/x86_64-pokysdk-linux/usr/lib/mipsel-poky-linux/gcc/mipsel-poky-linux/5.4.0/include-fixed",
		"/opt/dl-distro-tiny/2.2.4/sysroots/mips32r2el-poky-linux-musl/usr/include"
    ],
    "C_Cpp.default.compilerPath": "/opt/dl-distro-tiny/2.2.4/sysroots/x86_64-pokysdk-linux/usr/bin/mipsel-poky-linux-musl/mipsel-poky-linux-musl-g++",
    "C_Cpp.default.compileCommands": "${workspaceRoot}/build/compile_commands.json",
}
```

What do you suggest? (Apart from removing compileCommands, I'm already trying it :) )

**Comment by bobbrow** [comment]
If the configurationProvider setting is set, then cpptools doesn't use your compile_commands.json unless CMake Tools fails to provide the config for a given source file.

That's a really long issue you've referred to.  Can you summarize what's not working so we know what to focus on?

**Comment by gentooise** [comment]
> If the configurationProvider setting is set, then cpptools doesn't use your compile_commands.json unless CMake Tools fails to provide the config for a given source file.

I understand. But since both `compile_commands.json` and `configurationProvider` are generated by CMake Tools, which are the cases when a config is not in `configurationProvider` while is present in the `compile_commands.json`?

> That's a really long issue you've referred to. Can you summarize what's not working so we know what to focus on?

That was just for reference in case you needed more info about my configuration, toolchain, etc.
Ignore it if you don't need it.

**Comment by bobbrow** [comment]
I should have been more clear.  There shouldn't be a case where `compile_commands.json` has information that the configuration provider does not.  What I meant to say is that the other properties in `c_cpp_properties.json` are ignored when the configuration provider successfully provides a configuration.  If it does not, the cpptools will look in the `compile_commands.json` file next.  It probably won't find the info there either, so as a last resort it will try the `includePath` and `defines` set in the configuration.

**Comment by andreeis** [comment]
@gentooise , if everything works fine for you now, we would like to close this issue. Let us know if you'd like us to investigate or fix anything else.

**Comment by gentooise** [comment]
It seems everything works fine with `configurationProvider` and without `compileCommands`.

You may close this. Thank you.

**Comment by andreeis** [comment]
Wonderful.
